Add particle list with field tags

streeve opened the pull request:
Wraps the AoSoA with more user friendly type tags and enables a thread-level particle interface. Adapted from Picasso

Currently single AoSoA only, with a separate version in Cajita that stores the mesh as well.


Should add an example before merging

sslattery commented on core/src/Cabana_ParticleList.hpp:
```diff
@@ -0,0 +1,249 @@
+}
+
+//---------------------------------------------------------------------------//
+//! Return a single SoA for a single particle.
```
This returns a single element of a single field via `indices`

streeve replied:
Fixed

sslattery commented on core/src/Cabana_ParticleList.hpp:
```diff
@@ -0,0 +1,249 @@
+        particle.tuple(), indices... );
+}
+
+//! Return a single field for a single particle.
```
This returns a single element of a single field via `indices`

sslattery commented on core/src/Cabana_ParticleList.hpp:
```diff
@@ -0,0 +1,249 @@
+        particle.soa(), particle.vectorIndex(), indices... );
+}
+
+//! Return a single SoA for a single particle.
```
This returns a single element of a single field via `indices`

sslattery commented on core/src/Cabana_ParticleList.hpp:
```diff
@@ -0,0 +1,249 @@
+//---------------------------------------------------------------------------//
+// Particle accessors.
+//---------------------------------------------------------------------------//
+//! Return a single field for a single particle.
```
This returns a single element of a single field via `indices`

sslattery commented on core/src/Cabana_ParticleList.hpp:
```diff
@@ -0,0 +1,249 @@
+    const aosoa_type& aosoa() const { return _aosoa; }
+
+    //! Get the label.
+    const std::string& label() { return _label; }
```
should only need to `const` version of this function because you always return a `const &`

streeve replied:
Fixed

sslattery commented on core/src/Cabana_ParticleList.hpp:
```diff
@@ -0,0 +1,249 @@
+    std::size_t size() const { return _aosoa.size(); }
+
+    //! Get the AoSoA.
+    aosoa_type& aosoa() { return _aosoa; }
```
Did I have some functions that also allowed one to get a `Particle` or `ParticleView` on device given a 1D particle index?

streeve replied:
I added another, but maybe `get`/`set` is confusing with the existing free functions

streeve replied:
Discussed in the next comments

sslattery commented on core/unit_test/tstParticleList.hpp:
```diff
@@ -0,0 +1,227 @@
+    Kokkos::parallel_for(
+        "modify", Kokkos::RangePolicy<TEST_EXECSPACE>( 0, num_p ),
+        KOKKOS_LAMBDA( const int p ) {
+            typename list_type::particle_type particle( aosoa.getTuple( p ) );
```
Yeah I'd like some helper functions to do this on device. We should be able to pass the particle list to the device and get the AoSoA out of it there.

streeve replied:
I added these, but maybe `get`/`set` is confusing with the existing free functions

sslattery commented on core/unit_test/tstParticleList.hpp:
```diff
@@ -0,0 +1,227 @@
+    Kokkos::parallel_for(
+        "modify", Kokkos::RangePolicy<TEST_EXECSPACE>( 0, num_p ),
+        KOKKOS_LAMBDA( const int p ) {
+            auto s = Cabana::Impl::Index<
```
Definitely want a helper function for this

streeve replied:
I added, but maybe `access` isn't the best name

streeve commented on core/src/Cabana_Fields.hpp:
```diff
@@ -0,0 +1,167 @@
+//---------------------------------------------------------------------------//
+//! Particle position field type.
+template <std::size_t NumSpaceDim>
+struct Position : Vector<double, NumSpaceDim>
```
Removed most, but left `Position` to discuss if we want to use it in algorithms

sslattery commented on core/src/Cabana_ParticleList.hpp:
```diff
@@ -0,0 +1,270 @@
+
+    //! Get a single particle.
+    template <class IndexType>
+    KOKKOS_INLINE_FUNCTION auto get( IndexType p ) const
```
let's call this `getParticle()`

sslattery commented on core/src/Cabana_ParticleList.hpp:
```diff
@@ -0,0 +1,270 @@
+
+    //! Set a single particle.
+    template <class ParticleType, class IndexType>
+    KOKKOS_INLINE_FUNCTION void set( ParticleType particle, IndexType p ) const
```
lets call this `setParticle()`

sslattery commented on core/src/Cabana_ParticleList.hpp:
```diff
@@ -0,0 +1,270 @@
+
+    //! Set a single particle with the struct+array indexing.
+    template <class IndexType>
+    KOKKOS_INLINE_FUNCTION auto access( IndexType p ) const
```
lets call this `getParticleView()`

sslattery reviewed (changes requested):
Just rename those functions and this is good to merge.

streeve commented:
@sslattery renamed functions, cleaned up history, and removed the internal field examples like you previously suggested (we can add if we ever want them in an internal algorithm)

Tested with HIP manually
